### PR TITLE
Implement code boundaries

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- Implement suggested changes to CODEOWNERS file as described in #8525 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by HotThoughts** [comment]
@virtualroot could you add `@RasaHQ/infrastructure` and `@RasaHQ/research` to this repo? Looks like the same situation we have for rasa-sdk earlier. We need to grant write access to them so they can get assigned. 

> The people you choose as code owners must have write permissions for the repository. When the code owner is a team, that team must have write permissions, even if all the individual members of the team already have write permissions directly, through organization membership, or through another team membership.
> 
> https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/about-code-owners

**Comment by virtualroot** [comment]
> @virtualroot could you add `@RasaHQ/infrastructure` and `@RasaHQ/research` to this repo? Looks like the same situation we have for rasa-sdk earlier. We need to grant write access to them so they can get assigned.

✅ Done, `@RasaHQ/infrastructure` and `@RasaHQ/research` added with write access.

File: CODEOWNERS
Comment (wochinge) [inline CODEOWNERS:9]:
```suggestion
/rasa/core/featurizers/ @RasaHQ/research
```

File: CODEOWNERS
Comment (wochinge) [inline CODEOWNERS:22]: How about simplifying this and doing something like the following (according to the GitHub docs the **last match** takes precedence)
```suggestion
/rasa/core/policies/*.py @RasaHQ/research
/rasa/core/policies/ensemble.py @RasaHQ/enable-squad
/rasa/core/policies/registry.py @RasaHQ/enable-squad
```

File: CODEOWNERS
Comment (wochinge) [inline CODEOWNERS:40]: this seems to be mostly enable-squad - what about making enable-squad default for `rasa/core/*` and then just adding the exceptions for research?

**Review by wochinge — COMMENTED** [review]
Thanks for taking this one 💯 What do you think of adding a default rule which states that every python file should be reviewed by `@enable-squad` and then only adding the exception for research? This would in my opinion have the following advantages:
* if somebody creates a new file, it would automatically have at least some ownership. If think this doesn't make sense for us to review we can still change the code owners file but nothing falls through the cracks
* hopefully shorter and easier to read `CODEOWNERS` file?

File: CODEOWNERS
Comment (ancalita) [reply CODEOWNERS:40]: Sure, I'll action that - how do I make sure that a certain file has no codeowners, would this work?
`! /rasa/core/constants.py @RasaHQ/enable-squad`

**Comment by ancalita** [comment]
> What do you think of adding a default rule which states that every python file should be reviewed by `@enable-squad` and then only adding the exception for research

Sure - would this be something like this `*.py @RasaHQ/enable-squad` added right at the top after L4?

File: CODEOWNERS
Comment (wochinge) [reply CODEOWNERS]: oh, very good catch! Should we just try out? 🙈

File: CODEOWNERS
Comment (wochinge) [inline CODEOWNERS:7]: do we even need these two then?

File: CODEOWNERS
Comment (wochinge) [inline CODEOWNERS:5]: I think we've to limit this to the `rasa/` dir (otherwise we also get assigned for tests). According to the docs it would be something like this:
```suggestion
rasa/**.py @RasaHQ/enable-squad
```

File: CODEOWNERS
Comment (wochinge) [reply CODEOWNERS:7]: (also some of them below seem to become superfluous I think 🤔

**Review by wochinge — COMMENTED** [review]
I think we should be able to test it by just creating PRs against this branch, no?

File: CODEOWNERS
Comment (ancalita) [reply CODEOWNERS]: Alright, I'm thinking I should create branches for 4 distinct changes in: 
-  Enable owned file
-  file without owner
-  Research owned file
- dual Enable-Research ownership
I'll make some minimal change (add a comment this is for testing CODEOWNERS so it acts as a heads-up 😄 )

Comment (ancalita) [reply CODEOWNERS]: Hmm, first test didn't work out :( I created [this test PR](https://github.com/RasaHQ/rasa/pull/8932) but without the expected result. Not sure where things went wrong, should I target `main` instead?

Comment (ancalita) [reply CODEOWNERS]: I'm actually wondering if this syntax is valid: `/rasa/**.py @RasaHQ/enable-squad` - where did you find this in the docs? All I could find is the below - which actually states the opposite (that nested files won't be targeted):
```
# The `docs/*` pattern will match files like
# `docs/getting-started.md` but not further nested files like
# `docs/build-app/troubleshooting.md`.
docs/*  docs@example.com
```
Maybe the rule I should use is `rasa/ @RasaHQ/enable-squad` which will target **all** files in this directory as instructed here:
```
# In this example, @octocat owns any file in an apps directory
# anywhere in your repository.
apps/ @octocat
```

**Comment by ancalita** [comment]
> I think we should be able to test it by just creating PRs against this branch, no?

I think the reason why my test failed yesterday was also because this branch is not a [protected branch](https://docs.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-pull-request-reviews-before-merging). Probably the same will go for any `2.x` branches.
Who should I ask for this branch to be protected so I can test PRs against it?

**Comment by ancalita** [comment]
Hey @wochinge,

I've discovered another source of invalid syntax which was the `!` line, I've removed this and I've double-checked the shield icon on top of a few files and it shows the correct owner as described in the rules. The only disadvantage is that the `constants.py` file that shouldn't have any owners now has Enable as owner, but negating patterns is not yet possible in CODEOWNERS, as described in the docs (I was just reading it backwards at first 😅 ).

In addition, I also created a [test PR](https://github.com/RasaHQ/rasa/pull/8940) which tagged the correct code-owner.
To avoid spamming Research, do you think this is sufficient and ready to be merged? Or should I proceed opening 2 new PRs to test this?

File: CODEOWNERS
Comment (HotThoughts) [inline CODEOWNERS:1]: `/docs/docs/prototype-an-assistant.mdx @alwx`

Should we remove Rick as he left?

Comment (ancalita) [reply CODEOWNERS:1]: maybe I can tag the squad instead of individuals to avoid this situation when they leave?

File: CODEOWNERS
Comment (wochinge) [inline CODEOWNERS:5]: Should we do a double star to match also subdirectories in case they're created?
```suggestion
/rasa/core/policies/**.py @RasaHQ/research
```

Comment (wochinge) [reply CODEOWNERS:5]: or we just do 
```suggestion
/rasa/core/policies/ @RasaHQ/research
```

**Review by wochinge — APPROVED** [review]
> but negating patterns is not yet possible in CODEOWNERS, as described in the docs (I was just reading it backwards at first 😅 ).

The same happened to me 🙈 

> The only disadvantage is that the constants.py file that shouldn't have any owners now has Enable as owner, 

I think that's fine. I think the CODEOWNERs file is way more transparent the way it is now 💯 

Thanks for doing a test PR 💯

File: CODEOWNERS
Comment (ancalita) [reply CODEOWNERS:5]: Yup, the latter approach makes more sense 👍

**Pull request merged**
